blog/_posts/2017-06-25-dsl.md

```diff
@@ -0,0 +1,550 @@
+
+
+```julia
+(m::NaiveModel)(x) = m.f(x)
+```
+
+so that we can now just write
+
+
+```julia
+m1(10)
+```
+
+
+
+
+    20
+
+
+
+## Parametrising the type
+
+Since `Function` is an abstract type, for performance we should *not* have a field of this type inside our object.
+Rather, we parametrise the type using the type of the function:
+
+
+```julia
+struct Model{F}
+    f::F
+end
+
+(m::Model)(x) = m.f(x)
+```
+
+
+```julia
+m2 = Model(x->2x)
+```
+
+
+
+
+    Model{##3#4}(#3)
+
+
+
+
+```julia
+m2(10)
+```
+
+
+
+
+    20
+
+
+
+Let's compare the performance:
+
+
+```julia
+using BenchmarkTools
+```
+
+
+```julia
+@btime m1(10)
+```
+
+      41.482 ns (0 allocations: 0 bytes)
+
+
+
+
+
+    20
+
+
+
+
+```julia
+@btime m2(10)
+```
+
+      20.212 ns (0 allocations: 0 bytes)
+
+
+
+
+
+    20
+
+
+
+Indeed we have removed some overhead in the second case.
+
+## Defining a macro to create objects
+
+We now wish to define a macro that will allow us to use a simple syntax (that uses standard Julia syntax but itself is not valid Julia code) to create objects. In this case, we wish to write
+
+    @model 2x
+
+to define a `Model` object with the function `x->2x`.
+
+We will build up to macros by first building the tools (standard Julia functions) to manipulate the expression `2x` in the correct way to build a `Model` object from it.
+
+First, let's create a function to manipulate our expression:
+
+
+```julia
+function make_function(ex::Expr)
+    return :(x -> $ex)
+end
+```
+
+
+
+
+    make_function (generic function with 1 method)
+
+
+
+
+```julia
+ex = :(2x)
+make_function(ex)
+```
+
+
+
+
+    :(x->begin  # In[12], line 2:
+                2x
+            end)
+
+
+
+This function assumes that `ex` is an expression containing the variable `x` and makes a new expression representing an
+anonymous function with the single argument `x`. (See [my JuliaCon 2017 tutorial](https://github.com/dpsanders/julia_towards_1.0/blob/master/4.%20Metaprogramming.ipynb) for an example of how to walk through the expression tree in order to extract *automatically* the variables that it contains.)
+
+Now let's define a function `make_model` that takes a function and wraps it and passes it into a `Model` object:
+
+
+```julia
+function make_model(ex::Expr)
+    return :(Model($ex))
+end
+```
+
+
+
+
+    make_model (generic function with 1 method)
+
+
+
+
+```julia
+make_model(make_function(:(2x)))
+```
+
+
+
+
+    :(Model((x->begin  # In[12], line 2:
+                    2x
+                end)))
+
+
+
+If we evaluate this "by hand", we see that it correctly creates a `Model` object:
+
+
+```julia
+m3 = eval(make_model(make_function(:(2x))))
+```
+
+
+
+
+    Model{##7#8}(#7)
+
+
+
+
+```julia
+m3(10)
+```
+
+
+
+
+    20
+
+
+
+## Making a macro
+
+However, this is ugly and clumsy. Instead, we now wrap everything inside a **macro**.
+
+A **macro** is a kind of "super-function" that manipulates code.
```

Comment:
Calling a macro as a kind of function does not sound right, technically - although it may convey the right idea for the first time reader.